gluten-core/src/main/scala/io/glutenproject/extension/ColumnarPullOutProject.scala

```diff
@@ -0,0 +1,183 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.glutenproject.extension
+
+import io.glutenproject.execution._
+import io.glutenproject.utils.PullOutProjectHelper
+
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.expressions.aggregate.AggregateFunction
+import org.apache.spark.sql.catalyst.expressions.aggregate.Partial
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.catalyst.trees.TreeNodeTag
+import org.apache.spark.sql.execution.SparkPlan
+
+import scala.collection.mutable.ListBuffer
+
+trait ProjectProcessedHint
+case class PROJECT_PROCESSED() extends ProjectProcessedHint
+
+object ProjectProcessedHint {
+  val TAG: TreeNodeTag[ProjectProcessedHint] =
+    TreeNodeTag[ProjectProcessedHint]("io.glutenproject.projectprocessedhint")
+
+  def isAlreadyTagged(plan: SparkPlan): Boolean = {
+    plan.getTagValue(TAG).isDefined
+  }
+
+  def isPostProjectProcessed(plan: SparkPlan): Boolean = {
+    plan
+      .getTagValue(TAG)
+      .isDefined && plan.getTagValue(TAG).get.isInstanceOf[PROJECT_PROCESSED]
+  }
+
+  def postProjectProcessDone(plan: SparkPlan): Unit = {
+    tag(plan, PROJECT_PROCESSED())
+  }
+
+  private def tag(plan: SparkPlan, hint: ProjectProcessedHint): Unit = {
+    plan.setTagValue(TAG, hint)
+  }
+}
+
+object ColumnarPullOutPostProject extends Rule[SparkPlan] with PullOutProjectHelper {
+
+  override def apply(plan: SparkPlan): SparkPlan = plan.transform {
+    case agg: HashAggregateExecBaseTransformer
+        if !ProjectProcessedHint.isPostProjectProcessed(agg) &&
```

Comment:
why do we need this hint ? after we pull out post project, it should nothing happen when we go in this rule next time

Reply:
It can be removed here. In a previous version of implementation, I did not rewrite the `resultExpressions` of agg, which caused it to always match here. I will remove this check in agg.

However, the hint might still be needed for implementing join because join requires a post-project when the order of left and right need to be changed. This condition will always be satisfied for join.

Reply:
I'm not sure it is required for join, at least can add it when needed